packages/next-centra-checkout/src/createCentraProductsCacheServerHelpers.ts

```diff
@@ -0,0 +1,276 @@
+    const { segmentName } = options
+    const { ids: idsParam } = req.query
+    const segments = req.query[segmentName]
+
+    assert(typeof idsParam === 'string', "search parameter 'ids' should be a string")
+    assert(Array.isArray(segments), "search parameter 'segments' should be an array")
+
+    const [market, pricelist, language] = segments
+
+    assert(market, "missing search parameter 'market'")
+    assert(pricelist, "missing search parameter 'pricelist'")
+    assert(language, "missing search parameter 'language'")
+
+    const ids = idsParam.split(',')
+
+    const products = await callback({ market, pricelist, language }, ids)
+
+    if (options.responseHeaders) {
+      res.setHeaders(options.responseHeaders)
+    }
+
+    res.status(200).json(products)
+  }
+}
+
+/**
+ * Define Centra product cache server helpers.
+ * Helpers to be called server-side to retrieve cached data from Centra.
+ * To be used in conjunction with `createCentraProductCacheClientHelpers`
+ */
+export function createCentraProductCacheServerHelpers<TProduct extends ProductBase = ProductBase>(
+  options: ServerHelperOptions<TProduct>,
+) {
+  const {
+    cache,
+    centraCheckout: centraCheckoutCredentials,
+    getSessionContexts,
+    formatProduct,
+  } = options
+
+  return {
+    /**
+     * Fetch products, using the Cache-Aside pattern.
+     * Will primarily get products from the cache.
+     * If one or more products were not resolved from cache, those products will be retrieved from
+     * the Centra Checkout API.
+     *
+     * @param sessionContext - The session context to retrieve the products
+     *
+     * @see {@link https://learn.microsoft.com/en-us/azure/architecture/patterns/cache-aside}
+     */
+    fetchProducts: async (sessionContext: SessionContext, ids: string[]) => {
+      const configuredSessionContexts = await getSessionContexts()
+
+      const isSessionContextConfigured = Boolean(
+        configuredSessionContexts.find(
+          (configuredSessionContext) =>
+            configuredSessionContext.market === sessionContext.market &&
+            configuredSessionContext.pricelist === sessionContext.pricelist &&
+            configuredSessionContext.language === sessionContext.language,
+        ),
+      )
+
+      const compareIdsByInputOrder = (a: string, b: string) => {
+        const aIdx = ids.findIndex((id) => id === a)
+        const bIdx = ids.findIndex((id) => id === b)
+
+        return aIdx - bIdx
+      }
+
+      if (!isSessionContextConfigured) {
+        // Since the passed session context is not known among the session contexts, we'll try to
+        // fetch to perform the products request directly to source.
+        // This is due to that we will not update this session context in our `updateProducts`
+        // handler. If we would "Cache-Aside" these products, then we'll serve stale product data.
+        return fetchProductsCentra<TProduct>(centraCheckoutCredentials, sessionContext, ids)
```

Comment:
This is invalid.
The `TProduct` type is the one that should be returned from the exposed `fetchProducts`, not the internal`fetchProductsCentra` function.